Improve reindex

abi87 opened the pull request:
Following on https://github.com/HorizenOfficial/zen/issues/347 I introduced a `-reindexfast` option which allows to skip some expensive checks for txes in blocks below checkpoints, upon loading files from disk.
The mechanism is exactly that following upon receiving blocks from the network, header first

cronicc committed:
Merge pull request #335 from HorizenOfficial/development

abi87 committed:
Minor renaming before -reindex-fast implementation

abi87 committed:
Added UT file for reindex

abi87 committed:
Added UTs for LoadBlocksFromExternalFile

abi87 committed:
reindexU UTs: added UT to store and retrieve genesis to MapBlockIndex

abi87 committed:
Minor refactoring for readability

abi87 committed:
Reindex UTs: tested case for multiple blocks per file + minor refactoring of CreateNewBlock

abi87 committed:
Refactored generation of equihash for blockHeaders

abi87 committed:
Extended asserts in reindex UTs

abi87 committed:
reindex: refactored blocks loading from LoadBlocksFromExternalFile

abi87 committed:
Reindex: minor fixes to LoadBlocksFromExternalFile

abi87 committed:
Reindex: minor fix of function signature

abi87 committed:
Draft of introduction of header-first block load

abi87 committed:
Merge branch 'improve_reindex' of https://github.com/ZencashOfficial/zen into improve_reindex

abi87 committed:
reindex: Added Uts for out-of-order blocks

abi87 committed:
reindex: added UTs for duplicated blocks in files

abi87 committed:
Minimal changes for reindex-fast: removed refactoring of blocks loading

abi87 committed:
reindex: fixed block processing following headers one

abi87 committed:
Added UT for headers and block back-to-back processing

abi87 committed:
reindex-fast: added global option to cli

alsala commented on src/init.cpp:
```diff
@@ -366,7 +366,8 @@ std::string HelpMessage(HelpMessageMode mode)
             "(default: 0 = disable pruning blocks, >%u = target size in MiB to use for block files)"), MIN_DISK_SPACE_FOR_BLOCK_FILES / 1024 / 1024));
     strUsage += HelpMessageOpt("-reindex", _("Rebuild block chain index from current blk000??.dat files on startup"));
-#if !defined(WIN32)
+    strUsage += HelpMessageOpt("-reindexfast", _("Rebuild block chain index from current blk000??.dat files on startup, skipping expensive checks for blocks below checkpoints"));
```
Should they be mutually exclusive? In that case specify it in help too

abi87 replied:
Indeed `-reindexfast` absorbs `-reindex`, so if you specify both reindex fast is performed.
However this behaviour should not be allowed, so I am going to forbid specifying them both

alsala commented on src/init.cpp:
```diff
@@ -1393,6 +1416,7 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
     // ********************************************************* Step 7: load block chain
 
     fReindex = GetBoolArg("-reindex", false);
+    fReindexFast = GetBoolArg("-reindexfast", false);
```
Should they be mutually exclusive? In that case handle the clash if any

abi87 replied:
agree, introduce check for mutual exclusivity

alsala commented on src/init.cpp:
```diff
@@ -1727,7 +1752,7 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
         LogPrintf("Unsetting NODE_NETWORK on prune mode\n");
         nLocalServices &= ~NODE_NETWORK;
-        if (!fReindex) {
+        if (!fReindex || !fReindexFast) {
```
Shouldn't it be &&?

abi87 replied:
agree, fixed

abi87 committed:
Fixes following code review

abi87 committed:
Fixed CheckBlockIndex for regtest checks

Commit:
Improved the processing of blocks when zk-SNARKs verification is not needed (e.g. when using checkpoints).

- Skip cryptographic objects initialization for PHGR transactions when no verification is required.
- Skip verification of GROTH transactions when possible.

Commit:
Fixed the creation of new blk??????.dat files.

- Fixed a bug that was causing the vinfoBlockFile vector not to
correctly add new elements when the creation of a new blk??????.dat file
was needed.

- Added a unit test for the fix.

Commit:
Merge remote-tracking branch 'origin/pt/proof_verifier_improvement' into improve_reindex

lander86 committed:
freezing the b2 version and fixing pigz command on MacOs

lander86 committed:
fix ipfs image url

lander86 commented on src/gtest/test_reindex.cpp:
```diff
@@ -0,0 +1,604 @@
+    CCriticalBlock           csMainLock;        /**< Critical section needed when compiled with --enable-debug, which activates ASSERT_HELD */
+};
+
+TEST_F(ReindexTestSuite, LoadingBlocksFromNullFilePtrWillAbort) {
```
This test is not portable and makes the compilation on windows fail.

Commit:
Removed a minor unit test that could not be compiled for Windows target.

lander86 committed:
fix indentation and minor fixes

Commit:
Merge branch 'lv/fix_ci' into improve_reindex

abi87 committed:
Reverted minor change over readability disagreement
